aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -185,21 +185,32 @@ Tensor _fft_cufft(const Tensor& self, int64_t signal_ndim,
   std::vector<long long int> inembed(signal_ndim);
 
   // check the input sizes and strides to see if we need to make it contiguous
-  bool need_contiguous = input.stride(signal_ndim) == 0;
+  // cuFFT doesn't support batch dim with stride 0
+  bool need_contiguous = input.stride(0) == 0;
+
   if (complex_input) {
-    // Real/imag dimension must be like complex type. Need to make the input
-    // tensor contiguous if this dimension is not contiguous.
+    // Real/imag dimension must be like complex type.
     need_contiguous |= input.stride(-1) != 1;
+    // Strides of other dimensions needs to be aligned when viewed as of
+    // complex type, i.e., multiples of 2. We check the batch dim here. The
+    // other signal dims are checked in a for-loop afterwards.
+    need_contiguous |= input.stride(0) % 2 != 0;
   } else if (is_half) {
     // For half, base strides on the real part of real-to-complex and
     // complex-to-real transforms are not supported. Since our output is always
     // contiguous, only need to check real-to-complex case.
     need_contiguous |= input.stride(signal_ndim) != 1;
   }
-  // store last tensor stride to infer inembed array
-  // complex input's last dim is size 2 and contiguous
+
+  // Store last dimension stride to infer inembed array
+  // This is used when `need_contiguous=False`, so we can assume that the last
+  // dimension is indeed aligned. Complex input's last signal dim can be viewed
+  // as having stride=2, where the unit is sizeof(real_type).
   long long int ilast_stride = complex_input ? 2 : 1;
-   // for each signal dim from innermost to outermost
+  // For each signal dim from innermost to outermost, compute inembed values if
+  // possible. If not, then we need to make input contiguous. Notice that if we
+  // can compute inembed values, it means that the dimensions conform to the
+  // data type alignment, e.g., strides are multiples of 2 when complex_input.
   for (int64_t i = signal_ndim - 1; i >= 0; i--) {
```

Comment:
This is the for loop, right? I don't see a test for alignment in this loop.

FWIW, I'm patterning off of

```
+    for (int64_t i = 0; !need_contiguous && i <= signal_ndim; i++) {
+      need_contiguous |= input.stride(i) % 2 != 0;
+    }
```

which is obviously an alignment test. Is there a non-obvious alignment test in this loop? Worth calling out with a comment...

Reply:
Yeah... i have it in the commit right above.
```
 Notice that if we
+  // can compute inembed values, it means that the dimensions conform to the
+  // data type alignment, e.g., strides are multiples of 2 when complex_input.
```

Reply:
Aha! I see it!

Now that I think I actually understand what this code is doing (as opposed to just doing "spot checks" on it), I have some comments about how it can be made clearer (unless, @colesbury, you think this code is perfectly clear as it stands? :o)

1. Let's actually inline an explanation of what `inembeds` and `onembeds` are. At the moment, in this file, it seems the most explanation it gets is that they are "the sizes of enclosing tensor." The Advanced Data Layout section of the cufft has a very poorly written section http://docs.nvidia.com/cuda/cufft/index.html#advanced-data-layout but the equations (e.g., `input[b * idist + (x * inembed[1] + y) * istride]`) convey the important information from which you can work out the rest. (Actually, that's not true; the docs add "Note that the size of each dimension of the transform should be less than or equal to the inembed and onembed values for the corresponding dimension, that is n[i] ≤ inembed[i], n[i] ≤ onembed[i]"; which means you can't pass broadcasted tensors (stride == 0) to cufft; judging that the mkl variant doesn't do this test, I'm guessing it doesn't have a similar restriction?

2. Once this explanation exists, there will be multiple places where a code reader might see `inembed` or `onembeds` and wonder what it is. If you make the explanation in (1) a Note and reference it from each relevant call site, you can make it more likely that someone will actually write the comment you lovingly crafted in (1).

3. In fact, it would be nice to have a general discussion about weird stride modes, and possibly factor out some of the logic in this function. Not for this PR, I've filed an issue for it. https://github.com/pytorch/pytorch/issues/6165

4. What you might consider for this PR, from above, is refactoring the inembed calculation into its own function. The signature looks like `at::optional<std::vector<int64_t>> is_embedded_strides(std::vector<int64_t> stride)`. This returns `std::nullopt` when strides are not embedded (as per the cufft definition, I suppose, which means you ought to test for broadcasting too) and returning the multiplicative relationship between the strides when they are embedded.

Thanks for reading! Let me know if you disagree with any of this.

Reply:
1 & 2. I will write an explanation of `inembed` and `oembed`. Thanks! :) Yes, MKL doesn't have such restrictions.
4. Sounds reasonable. Should this function live in this file, or in maybe `TensorUtils.h`?

Reply:
offline convo conclusion: put it in the file